Parallel SparseVector and SparseMatrix operations

artem-burashnikov opened the pull request:
## Задачи

> 1. Научиться читать файлы в формате matrixmarket.
> 2. Научиться строить граф на основе прочитанной матрицы смежности. 
> 3. Реализовать параллельные версии матричных и векторных операций.

artem-burashnikov committed:
Add reading MM file format for real-valued matrices

artem-burashnikov committed:
Fix unable to build: remove new keyword

artem-burashnikov committed:
Test try use and new keywords

gsvgit commented on src/spbu-fsharp/MatrixReader.fs:
```diff
@@ -7,7 +7,7 @@ open SparseMatrix.SparseMatrix
 
     // Read the first line which contains general information about a given file in a MM format.
-    let reader = StreamReader(filePath)
+    use reader = new StreamReader(filePath)
```
А чем readalllines не угодил?

artem-burashnikov replied:
Кажется, что это дело будет потом менее удобно обрабатывать, ведь в шапке не фиксированное число комментариев, судя по спецификации формата.

gsvgit replied:
Ща... Формат файла в любом случае строковый. Прочитали строки, а потом сколько надо, столько и взяли.

artem-burashnikov committed:
Test disable fsharplint rule for the specific line

artem-burashnikov committed:
Change MatrixReader now uses ReadAllLines method

artem-burashnikov committed:
Remove redundant code

artem-burashnikov committed:
Add pre-commit git hook

gsvgit commented on src/spbu-fsharp/MatrixReader.fs:
```diff
@@ -8,22 +8,22 @@ let readMatrixToSparse (filePath: string) =
+    let symmetry = metaData[4]
+
+    // Skip all comments and read matrix parameters and data.
     let mutable dataLineIndex = 1
```
А может можно просто [Seq.skipWhile](https://fsharp.github.io/fsharp-core-docs/reference/fsharp-collections-seqmodule.html#skipWhile)?

artem-burashnikov committed:
Change everything

artem-burashnikov committed:
Add Fail-safe check for file-reading and accessing arrays

artem-burashnikov committed:
Fix static member name

artem-burashnikov committed:
Change symmetry types, field types
Add MatrixReader reading methods for real, integer, pattern fields

artem-burashnikov committed:
Fix formatting
Fix git hook

artem-burashnikov commented:
- [ ] Реализовать чтение файлов в формате matrix market.

artem-burashnikov committed:
Add mirror of matrix coordinates of symmetric data

artem-burashnikov commented:
- [ ] Научиться читать файлы в формате matrixmarket.

artem-burashnikov committed:
Change consistent timing of SeqToList conversion accross data match patterns

artem-burashnikov commented:
- [ ] Научиться читать файлы в формате matrixmarket.

artem-burashnikov committed:
Change more detailed failwith messages

artem-burashnikov committed:
Remove unsupported symmetry types

artem-burashnikov committed:
Update build.yml

artem-burashnikov committed:
Update build.yml

artem-burashnikov committed:
Update build.yml

artem-burashnikov committed:
Change bool to unit value in pattern type matricies

artem-burashnikov committed:
Remove MatrixReader redundant code
Add Graphs initial idea

artem-burashnikov committed:
Add .fantomasignore file to fix the fantomas git hook
Change vertex, edge, graph types

artem-burashnikov committed:
Change vertex, edge, graph type annotations

artem-burashnikov committed:
Change graph constructor

artem-burashnikov committed:
Change edge type to undirected edge

artem-burashnikov committed:
Fix typo
Remove redundant code

artem-burashnikov committed:
Add paralell elementwise vector operations
Add parallel vector by matrix multiplication
Change tests to run parallel operations

artem-burashnikov committed:
Fix linter
Change fDo function arguments for clarity

artem-burashnikov committed:
Remove .fantomasignore
Remove redundant code while reading all lines from a given file

gsvgit commented on src/spbu-fsharp/MatrixAlgebra.fs:
```diff
@@ -50,11 +50,107 @@ module MatrixAlgebra =
+            BinTree.Node(results[0], results[1]) |> SparseVector.VectorData.reduce
+
+        /// Non-parallel workflow
+        let rec innerN bTree1 bTree2 =
```
Кажется, что-то такое было где-то выше.

artem-burashnikov replied:
Убрал.

gsvgit commented on src/spbu-fsharp/MatrixAlgebra.fs:
```diff
@@ -50,11 +50,107 @@ module MatrixAlgebra =
+
+        /// Parallel workflow
+        let rec innerP parallelLevel bTree1 bTree2 =
+            match bTree1, bTree2 with
```
И это очень похоже. Можно ли как-то не писать одну и ту же логику несколько раз?

artem-burashnikov replied:
> И это очень похоже. Можно ли как-то не писать одну и ту же логику несколько раз?

Вы про соединить innerN и innerP или про elementwiseVecVec и parallelElementwiseVecVec?

gsvgit replied:
Я про соединить всё, что можно соединить. Зачем дублировать код, если можно не дублировать?

artem-burashnikov replied:
Убрал.

gsvgit commented on src/spbu-fsharp/MatrixAlgebra.fs:
```diff
@@ -186,3 +274,192 @@ module MatrixAlgebra =
+
+
+    /// Vector by Matrix multiplication utilizing parallel computations.
+    let parallelVecByMtx
```
Ну и тут попахивает дублированием.

gsvgit commented on src/spbu-fsharp/MatrixReader.fs:
```diff
@@ -0,0 +1,180 @@
+        COOMatrix(data, file.Rows, file.Columns) |> SparseMatrix
+
+    member this.Pattern =
+        if file.Field <> Pattern then
```
С одной стороны, решение имеет место быть. С другой --- Вы не можете, например, поднять булев паттерн из интовой матрицы, что странно.

А ещё есть ощущение, что Вы эти методы как следует не тестировали.

gsvgit replied:
Ну да. Чот тестов на ридер как-то не сильно много)

artem-burashnikov replied:
Хардкодить путь к текстовому файлу для тестов - это насколько нормально?

artem-burashnikov replied:
Добавил некоторые тесты.

gsvgit commented on tests/spbu-fsharp.Tests/MatrixAlgebraTests.fs:
```diff
@@ -458,8 +458,8 @@ module Algebra =
                   let vec2 = SparseVector arr2
-                  let result1 = elementwiseVecVec fPlus vec1 vec2
-                  let result2 = elementwiseVecVec fPlus vec2 vec1
+                  let result1 = parallelElementwiseVecVec 0 fPlus vec1 vec2
```
А есдли не 0? Будет хоть что-то параллелиться?

artem-burashnikov replied:
> А есдли не 0? Будет хоть что-то параллелиться?

Вы намекаете, что сложение векторов неверно распараллелено?

gsvgit replied:
Ну.... Как минимум, странно,  что тестов нету)

artem-burashnikov replied:
Все, что не 0, работает заметно медленнее... В плане времени выполнения тестов.

artem-burashnikov replied:
> Ну.... Как минимум, странно, что тестов нету)

Все, что не 0, работает заметно медленнее... В плане времени выполнения тестов.

gsvgit replied:
Ага... Ну и вернёмся к Вашему же вопросу. А верно ли распараллелено тогда?

artem-burashnikov replied:
Ну... Это может быть связано с тем, как работают тесты, например.
Просто, смотря на код, кажется, что все правильно: есть левая ветка дерева, есть правая, которые считаются параллельно.

gsvgit replied:
Ну попробуйте что-то конкретное уже забенчмаркать)

artem-burashnikov replied:
Еще надо понять, как)

gsvgit replied:
Самое время. Всё равно следующий шаг --- эксперименты.

artem-burashnikov replied:
Вроде бы распараллелил. Посмотрите, пожалуйста.

artem-burashnikov committed:
Remove redundant code in MatrixAlgebra
Remove checking the pattern field of MatrixReader to be able to read any matrix as pattern
Change name of elementwiseVecVec function in MatrixAlgebra to vecMap2

artem-burashnikov committed:
Remove redundant code

artem-burashnikov committed:
Fix naming
Change computation levels to 0 for tests

artem-burashnikov committed:
Add BencmharkDotNet to the project

artem-burashnikov committed:
Change parallel vectorMap2 (maybe works?)
Change vectorMap2 test for parallel testing
Add primitive benchmarks

artem-burashnikov committed:
Change vectorMap2

artem-burashnikov committed:
Fix formatting

artem-burashnikov committed:
Test MatrixReader reading external files (hardcoded path)

artem-burashnikov committed:
Test __SOURCE_DIRECTORY__ in file path

artem-burashnikov committed:
Add float parser, integer parser. Change MatrixReader to use parsers

artem-burashnikov committed:
Fix parallel mult
Fix parallel add
Change Benchmarks

artem-burashnikov committed:
Fix formatting

artem-burashnikov committed:
Add MatrixReader tests

artem-burashnikov committed:
Change Benchmarks
Fix typo
Add latest bench

artem-burashnikov committed:
Fix formatting

artem-burashnikov committed:
Merge branch 'HomeWork7' into artem-burashnikov-patch-1

artem-burashnikov committed:
Squashed commit of the following:

commit b801b73a0e269d0cda4c6058ea9e43bf0595c1b9
Author: artem <artem.burashnikov@gmail.com>
Date:   Sat Feb 11 01:48:11 2023 +0300

    Remove redundant async wrappers in async calls

commit f5d87517c1f48657b859e627130338081917a6ee
Author: artem <artem.burashnikov@gmail.com>
Date:   Fri Feb 10 16:19:37 2023 +0300

    Refactor async vecByMtx calls

commit f86c2061c0be381887dce9ac7534bf764042b06e
Author: artem <artem.burashnikov@gmail.com>
Date:   Fri Feb 10 15:44:04 2023 +0300

    Fix typos in Benchmarks

commit 123e2b2d5c9538ce8264dfc19dd40cc9c299aee1
Author: artem <artem.burashnikov@gmail.com>
Date:   Fri Feb 10 15:23:48 2023 +0300

    Change vectorMap2 from function to static member of SparseVector type
    Refactor the code to adress the changes.

artem-burashnikov committed:
Add latest bench

artem-burashnikov commented:
Проблема была не в коде, который отвечал за распараллеливание, а в коде, который отвечал за бенчмарки.

gsvgit commented on src/spbu-fsharp/BenchmarkDotNet.Artifacts/results/MyBenchmarks.MatrixAlgebraBenchmarks.MyAddBench-report-github.md:
````diff
@@ -0,0 +1,15 @@
+```
+|    Method |    Mean |    Error |   StdDev | Ratio | RatioSD |
+|---------- |--------:|---------:|---------:|------:|--------:|
+|   BaseAdd | 1.394 s | 0.0274 s | 0.0346 s |  1.00 |    0.00 |
````
А куда делись уровни 0, 1, 3 ?

gsvgit commented on src/spbu-fsharp/Benchmarks.fs:
```diff
@@ -0,0 +1,61 @@
+            Array2D.init 8_000 8_000 (fun i j -> if (i + j) % 3 = 0 then Some(r.Next(1, 10)) else Option.None)
+            |> SparseMatrix
+
+        let vec =
```
А почему матрица `[<GlobalSetup>]`, а вектор --- нет?

artem-burashnikov replied:
Изменил весь код бенчмарка.

gsvgit commented on src/spbu-fsharp/Benchmarks.fs:
```diff
@@ -0,0 +1,61 @@
+            |> SparseVector
+
+        [<Benchmark(Baseline = true)>]
+        member this.BaseMult() = vecByMtx 0 fPlus fMult vec mtx
```
Ага. 0 --- вот он. А 1 и 3 всё ещё где-то потерялись.

gsvgit commented on src/spbu-fsharp/Graphs.fs:
```diff
@@ -0,0 +1,36 @@
+
+type Vertex<'A> = 'A
+
+type UndirectedEdge<'A when 'A: comparison> = Set<'A>
```
Лучше типвые переменные тоже как-то вменяемо называть. Это ведь `'Label`?

artem-burashnikov replied:
Думаю, что нет. Это имитация множества рёбер. Label это уже то, что на самом ребре.

artem-burashnikov replied:
``'Label`` мне кажется чем-то другим.

gsvgit commented on src/spbu-fsharp/Graphs.fs:
```diff
@@ -0,0 +1,36 @@
+    let edgesOfMtx (mtx: SparseMatrix<'A>) : Set<UndirectedEdge<uint>> =
+        Set.ofSeq (
+            seq {
+                for i = 1 to toIntConv mtx.Rows do
```
Ой. То есть вершины из разреженной матрицы всегда честно за квадрат?

artem-burashnikov replied:
Вершины линейно, ребра за квадрат. Думаю, ребра будет выгоднее обходом дерева собирать через какой-нибудь fold.

artem-burashnikov replied:
Почти такое ощущение, что тут с помощью BFS можно получить ответ.

artem-burashnikov replied:
Попытался эффективно.

gsvgit commented on src/spbu-fsharp/Helper.fs:
```diff
@@ -78,6 +78,17 @@ module Numbers =
+        try
+            float input
+        with _ ->
+            failwith "parseFloat: Invalid input"
```
Луче сразу показывать, что там за инпут неожиданный такой.

gsvgit commented on src/spbu-fsharp/MatrixReader.fs:
```diff
@@ -0,0 +1,182 @@
+    static member FormatFromStr str =
+        match str with
+        | "coordinate" -> Coordinate
+        | _ -> failwith "Unsupported format type"
```
Какой? Что там пришло?

gsvgit commented on src/spbu-fsharp/MatrixReader.fs:
```diff
@@ -0,0 +1,182 @@
+    static member ObjectFromStr str =
+        match str with
+        | "matrix" -> Matrix
+        | _ -> failwith "Unsupported object type"
```
Какой?

gsvgit commented on src/spbu-fsharp/MatrixReader.fs:
```diff
@@ -0,0 +1,182 @@
+
+    // Convert to sequence and skip all lines with comments.
+    // The first line after the last comment contains parameters of the matrix.
+    let sq = Array.toSeq allLines |> Seq.skipWhile (fun line -> line[0] = '%')
```
А может просто сразу seq читать? А то гоняете туд-обратно

artem-burashnikov replied:
Прочитал в массив, а потом один раз в sequence превратил. Дальше только sequence использую, кажется. ReadLines вроде бы читает в коллекцию, попробую.

gsvgit commented on src/spbu-fsharp/MatrixReader.fs:
```diff
@@ -0,0 +1,182 @@
+    let rows, columns, entries =
+        let head = Seq.tryHead sq
+
+        if head <> None then
```
Чот странное. Почему не длина просто?

artem-burashnikov replied:
Длину считать надо, а тут только один элемент проверить. Идея такая.

gsvgit commented on src/spbu-fsharp/MatrixReader.fs:
```diff
@@ -0,0 +1,182 @@
+            let i, j, v = triplet
+            (j, i, v)
+
+        Seq.skipWhile (fun (i, j, _) -> i = j) sq |> Seq.map mapping |> Seq.append sq
```
А правда ли, что диагональ всегда в начале?

artem-burashnikov replied:
Нет, это я плохо читаю описание того, что встроенные методы делают...

artem-burashnikov replied:
Не тот метод применил.

gsvgit commented on src/spbu-fsharp/MatrixReader.fs:
```diff
@@ -0,0 +1,182 @@
+        COOMatrix(data, file.Rows, file.Columns) |> SparseMatrix
+
+    member this.Integer =
+        if file.Field <> Integer then
```
А можно как-то без дублирования кода обойтись?

artem-burashnikov replied:
Вот тут не так просто... Могу только предложить одну функцию, которая будет принимать converter и возвращать помимо всего прочего converter stringValue. Но сделать функцию, которая вернёт или матрицу int, или матрицу float, или какую-то друую не знаю, как. Такого полиморфизма вроде бы и нету.

artem-burashnikov replied:
Только если сделать общую функцию, которая извне принимает какой-нибудь converter и возвращает, помимо всего прочего, ``converter valueInStringFormat``
Но тогда пользователем должен передаваться этот самый converter.

gsvgit commented on src/spbu-fsharp/SparseVector.fs:
```diff
@@ -183,3 +195,64 @@ module SparseVector =
+            if vec1.Length <> vec2.Length then
+                failwith "parallelElementwiseVecVec: Dimensions of objects don't match."
+            else
+                SparseVector(inner computationLevel vec1.Data vec2.Data, vec1.Length)
```
Так он `parallelLevel` или `computationLevel`?

artem-burashnikov replied:
``parallelLevel``

gsvgit commented on src/spbu-fsharp/SparseVector.fs:
```diff
@@ -183,3 +195,64 @@ module SparseVector =
 
         member this.IsEmpty = this.Data = BinTree.None
+
+        static member Map2
```
А как так получилось, что в дифах нет удаления старого кода про elementWise?

artem-burashnikov replied:
Я изменения сделал в отдельной ветке, которую создал от текущей, а потом ``git merge --squash`` обратно.
Вот коммит https://github.com/artem-burashnikov/2022-Programming-Technologies-F-sharp-course/commit/123e2b2d5c9538ce8264dfc19dd40cc9c299aee1

gsvgit replied:
А....

gsvgit commented on tests/spbu-fsharp.Tests/MatrixAlgebraTests.fs:
```diff
@@ -427,8 +427,8 @@ module Algebra =
                   let vecOnes = SparseVector arrOnes
-                  let plusOne = elementwiseVecVec fPlus vec vecOnes
-                  let plusMinusOne = elementwiseVecVec fMinus plusOne vecOnes
+                  let plusOne = SparseVector.Map2 0 fPlus vec vecOnes
```
А на больше чем 0 почему бы не потестить? Вообще как-то не вижу тестов на параллельные версии.

artem-burashnikov replied:
Не придумал, зачем тестировать эти функции выше, чем на уровне 0, если этот уровень сам по себе показывает корректность вычислений, а производительность показывают бенчмарки.

gsvgit replied:
А в друг Вы изменяемое состояние неправильно используете?

artem-burashnikov committed:
Change parseInt and parseFlot error output for clarity

artem-burashnikov committed:
Change name of parallel parameter in SparseVector.Map2 and MatrixAlgebra.vecByMtx
Change parallelLevel from int to uint

artem-burashnikov committed:
Change code to address paralellLevel type change from int to uint

artem-burashnikov committed:
Change parameter type name of UndidrectedEdge for clarity

artem-burashnikov committed:
Change reading file to array to reading file to sequence
Fix useSymmetry function
Remove separate MatrixReader reading methods
Add universal MatrixReader.Read that uses external converter as an input

artem-burashnikov committed:
Change MatrixReader tests to use external converter

artem-burashnikov committed:
Add parallel parameter tests

artem-burashnikov committed:
Change Benchmarks
Add parameters to Benchmarks
Add array and table initializers with density parameter

artem-burashnikov committed:
Add missing updated fsproj file

Change edgesOfMtx method to utilize tree structure for faster computation
Add edgesOfMtx test

artem-burashnikov committed:
Change better code reusability

artem-burashnikov committed:
Change Benchmarks parameters

gsvgit commented on src/spbu-fsharp/Graphs.fs:
```diff
@@ -0,0 +1,42 @@
+
+type Vertex<'A> = 'A
+
+type UndirectedEdge<'PairOfVertices when 'PairOfVertices: comparison> = Set<'PairOfVertices>
```
И при этом это не `Vertex<'A> * Vertex<'A>` Но как? И, главное, зачем?

artem-burashnikov replied:
Вообще, я, кажется, больше головной боли себе создал этими типами.

artem-burashnikov replied:
Хотел, чтобы было издали графоподобное.

gsvgit commented on src/spbu-fsharp/Graphs.fs:
```diff
@@ -0,0 +1,42 @@
+        )
+
+    let edgesOfMtx (mtx: SparseMatrix<'A>) : Set<UndirectedEdge<uint>> =
+        let mutable edges = Set.empty
```
Напишите уже общую ходилку по матрице типа `fold`, видимо.

gsvgit commented on src/spbu-fsharp/MatrixReader.fs:
```diff
@@ -0,0 +1,156 @@
+            let entries = uint size[2]
+            rows, columns, entries
+
+    let allLines = File.ReadLines filePath |> Seq.cast<string>
```
А `ReadLines` разве не строки читает? Зачем конвертация?

artem-burashnikov replied:
Он читает в более общий тип. Нужно или так делать, чтобы дальше работать с seq, или писать `seq { yield! ... }`, что, кажется, делает что-то подобное.

gsvgit replied:
Ээээ .. IEnumerable не сильно более общий)

artem-burashnikov replied:
С ним Seq.skipWhile не дружит. Во всяком случае, у меня не получалось подружить. Попробую потыкать ещё что, но...

artem-burashnikov replied:
> Ээээ .. IEnumerable не сильно более общий)

Ахаха, все магическим образом заработало без конвертации) Это, наверное, один из тех случаев, когда сначала нужно компилятору чуток дальше в коде подсказать, а потом раньше в коде можно ослабить условия.

gsvgit commented on tests/spbu-fsharp.Tests/GraphsTests.fs:
```diff
@@ -0,0 +1,44 @@
+                    for j = 0 to Array2D.length2 table - 1 do
+                        let value = table[i, j]
+
+                        if value <> Option.None then
```
`value.IsNone`

artem-burashnikov committed:
Change edgesOfMtx method to be similar to the general fold function. Remove redundant types for Vertex and Edge

artem-burashnikov committed:
Remove redundant casting to sequence after reading lines from a file

artem-burashnikov committed:
Change None-condition check to use built-in method for clarity and readability

artem-burashnikov committed:
Change fp variable (file path) to utilize Sysem.IO.Path.Combine

artem-burashnikov committed:
Change edgesOfMtx in Graphs to be a fold static method in SparseMatrix

artem-burashnikov committed:
Fix SparseMatrix.Fold

gsvgit commented on src/spbu-fsharp/Graphs.fs:
```diff
@@ -0,0 +1,25 @@
+
+    let edgesOfMtx (mtx: SparseMatrix<'A>) =
+
+        let folder i j _ set = Set.add (Set.ofList [ i; j ]) set
```
1. Если я правильно понимаю, то метки на рёбрах Вы не сохраняете.
2. Не вплне пнял, что вообще происходит. Почему рёбра --- это что-то типа множества множеств?

artem-burashnikov replied:
Да, только рёбра.
Я пытаюсь подогнать то, что происходит, под определение неориентированного графа.
То есть получить по матрице смежности множество неупорядоченных рёбер. А чтобы пара была неупорядоченной, я её засунул во множество.

gsvgit replied:
Что?)

artem-burashnikov replied:
Ну то есть у графа можно посмотреть его список вершин и его список неупорядоченных ребер.
Хотел реализовать эти две операции.

artem-burashnikov committed:
Change edgesOfMtx method

artem-burashnikov commented:
Теперь я собираю по матрице множество троек (вершина, вершина, значение).

artem-burashnikov committed:
Merge branch 'HomeWork7' into artem-burashnikov-patch-1

gsvgit commented on src/spbu-fsharp/Graphs.fs:
```diff
@@ -0,0 +1,25 @@
+
+    let edgesOfMtx (mtx: SparseMatrix<'A>) =
+
+        let folder i j value set = Set.add (i, j, Some value) set
```
1. А вот на ребре значение всегда ровно то, что в матрице храниться. Не надо его ещё раз в Option заварачивать.
2. А вот тут как раз было бы уместно тип "Ребро" сдеолать аккуратно.

artem-burashnikov committed:
Remove redundant Option for values on edges in edgesOfMtx method in Graph type

artem-burashnikov committed:
Add Vertex and Edge types

artem-burashnikov committed:
Fix tests for Graphs

gsvgit reviewed (approved):
Пойдёт!) Тепреь эксперименты и отчёт.

artem-burashnikov committed:
Merge branch 'HomeWork7' into artem-burashnikov-patch-1
